// carbonldp/.esm2015/Pointer/Pointer.js
import { ModelDecorator } from "../Model/ModelDecorator";
import { isObject } from "../Utils";
export const Pointer = {
    PROTOTYPE: {
        get $id() { return ""; },
    },
    isDecorated(object) {
        return ModelDecorator
            .hasPropertiesFrom(Pointer.PROTOTYPE, object);
    },
    decorate(object) {
        if (Pointer.isDecorated(object))
            return object;
        return ModelDecorator
            .definePropertiesFrom(Pointer.PROTOTYPE, object);
    },
    is(value) {
        return isObject(value)
            && Pointer.isDecorated(value);
    },
    create(data) {
        const clone = Object.assign({}, data);
        return Pointer.createFrom(clone);
    },
    createFrom(object) {
        return Pointer.decorate(object);
    },
    areEqual(pointer1, pointer2) {
        return pointer1.$id === pointer2.$id;
    },
    getIDs(pointers) {
        return pointers
            .map(pointer => pointer.$id);
    },
    getID(pointerOrIRI) {
        return isObject(pointerOrIRI) ? pointerOrIRI.$id : pointerOrIRI;
    },
};
