// sparqler/.esm5/patterns/triplePatterns/PropertyBuilder.js
import { cloneElement } from "../../data/utils";
import { PropertyToken } from "../../tokens/PropertyToken";
import { getPropertyToken } from "../paths/utils";
import { convertValue } from "../utils";
function _cloneContainer(container, propertyToken) {
    var properties = container.targetToken.properties.concat(propertyToken);
    var targetToken = cloneElement(container.targetToken, { properties: properties });
    return cloneElement(container, { targetToken: targetToken });
}
function _updateContainer(container, propertyToken) {
    container.targetToken.properties.push(propertyToken);
    return container;
}
function getHasFn(genericFactory, container) {
    return function (property, objects) {
        var verbToken = getPropertyToken(container, property);
        var propertyToken = new PropertyToken(verbToken);
        objects = Array.isArray(objects) ? objects : [objects];
        propertyToken.addObject.apply(propertyToken, objects.map(convertValue));
        var newContainer = container.targetToken.token === "subject" ?
            _cloneContainer(container, propertyToken) :
            _updateContainer(container, propertyToken);
        var genericObject = genericFactory(newContainer, {});
        return PropertyBuilderMore.createFrom(genericFactory, newContainer, genericObject);
    };
}
export var PropertyBuilder = {
    createFrom: function (genericFactory, container, object) {
        return Object.assign(object, {
            has: getHasFn(genericFactory, container),
        });
    }
};
export var PropertyBuilderMore = {
    createFrom: function (genericFactory, container, object) {
        return Object.assign(object, {
            and: getHasFn(genericFactory, container),
        });
    }
};
