// sparqler/.esm5/patterns/triplePatterns/RDFLiteral.js
import { cloneElement } from "../../data/utils";
import { LanguageToken } from "../../tokens/LanguageToken";
import { XSD } from "../../utils/XSD";
import { TripleSubject } from "./TripleSubject";
function getWithTypeFn(container) {
    return function (type) {
        if (type in XSD)
            type = XSD[type];
        var iriType = container.iriResolver.resolve(type, true);
        var subject = cloneElement(container.targetToken.subject, { type: iriType });
        var targetToken = cloneElement(container.targetToken, { subject: subject });
        var newContainer = cloneElement(container, { targetToken: targetToken });
        return TripleSubject.createFrom(newContainer, {});
    };
}
function getWithLanguageFn(container) {
    return function (language) {
        var langToken = new LanguageToken(language);
        var subject = cloneElement(container.targetToken.subject, { language: langToken });
        var targetToken = cloneElement(container.targetToken, { subject: subject });
        var newContainer = cloneElement(container, { targetToken: targetToken });
        return TripleSubject.createFrom(newContainer, {});
    };
}
export var RDFLiteral = {
    createFrom: function (container, object) {
        return TripleSubject.createFrom(container, Object.assign(object, {
            withType: getWithTypeFn(container),
            withLanguage: getWithLanguageFn(container),
        }));
    },
};
