// sparqler/.esm5/patterns/triplePatterns/TriplePatternsBuilder.js
import { Container } from "../../data/Container";
import { Factory } from "../../data/Factory";
import { BlankNodePropertyToken } from "../../tokens/BlankNodePropertyToken";
import { BlankNodeToken } from "../../tokens/BlankNodeToken";
import { CollectionToken } from "../../tokens/CollectionToken";
import { LiteralToken } from "../../tokens/LiteralToken";
import { RDFLiteralToken } from "../../tokens/RDFLiteralToken";
import { SubjectToken } from "../../tokens/SubjectToken";
import { VariableToken } from "../../tokens/VariableToken";
import { Pattern } from "../Pattern";
import { convertValue } from "../utils";
import { BlankNodeBuilder } from "./BlankNodeBuilder";
import { RDFLiteral } from "./RDFLiteral";
import { TripleSubject } from "./TripleSubject";
function _getPatternContainer(container, token) {
    return new Container({
        iriResolver: container.iriResolver,
        targetToken: new SubjectToken(token),
    });
}
function _getTripleSubject(container, token) {
    var patternContainer = _getPatternContainer(container, token);
    return TripleSubject.createFrom(patternContainer, {});
}
function _getNodeSubject(container, token) {
    var patternContainer = _getPatternContainer(container, token);
    return Factory.createFrom(TripleSubject.createFrom, Pattern.createFrom)(patternContainer, {});
}
function getResourceFn(container) {
    return function (iri) {
        var token = container.iriResolver.resolve(iri);
        return _getTripleSubject(container, token);
    };
}
function getVarFn(container) {
    return function (name) {
        var token = new VariableToken(name);
        return _getTripleSubject(container, token);
    };
}
function getLiteralFn(container) {
    return function (value) {
        if (typeof value !== "string") {
            var token_1 = new LiteralToken(value);
            return _getTripleSubject(container, token_1);
        }
        var token = new RDFLiteralToken(value);
        var patternContainer = _getPatternContainer(container, token);
        return RDFLiteral.createFrom(patternContainer, {});
    };
}
function getCollectionFn(container) {
    return function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        var _a;
        var token = (_a = new CollectionToken()).addObject.apply(_a, values.map(convertValue));
        return _getNodeSubject(container, token);
    };
}
function _getBlankNode(container, label) {
    if (label && !label.startsWith("_:"))
        label = "_:" + label;
    var token = new BlankNodeToken(label);
    return _getTripleSubject(container, token);
}
function _getBlankNodeProperty(container, builderFn) {
    var token = new BlankNodePropertyToken();
    var builderContainer = new Container({
        iriResolver: container.iriResolver,
        targetToken: token,
    });
    var builder = BlankNodeBuilder.createFrom(builderContainer, {});
    builderFn(builder);
    if (token.properties.length < 1)
        throw new Error("At least one property must be specified by the self builder.");
    return _getNodeSubject(container, token);
}
function getBlankNodeFn(container) {
    return function (labelOrBuilderFn) {
        if (typeof labelOrBuilderFn === "function")
            return _getBlankNodeProperty(container, labelOrBuilderFn);
        return _getBlankNode(container, labelOrBuilderFn);
    };
}
export var TriplePatternsBuilder = {
    createFrom: function (container, object) {
        return Object.assign(object, {
            resource: getResourceFn(container),
            var: getVarFn(container),
            literal: getLiteralFn(container),
            collection: getCollectionFn(container),
            blankNode: getBlankNodeFn(container),
        });
    },
};
