// sparqler/.esm2015/patterns/notTriplePatterns/GroupPattern.js
import { Container } from "../../data/Container";
import { GroupPatternToken } from "../../tokens/GroupPatternToken";
import { UnionPatternToken } from "../../tokens/UnionPatternToken";
import { NotTriplePattern } from "./NotTriplePattern";
import { UnionPattern } from "./UnionPattern";
function getUnionFn(container) {
    return patterns => {
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        const newGroupToken = new GroupPatternToken();
        newGroupToken.patterns.push(...patterns.map(x => x.getPattern()));
        const unionToken = new UnionPatternToken();
        unionToken.groupPatterns.push(container.targetToken, newGroupToken);
        const newContainer = new Container({
            iriResolver: container.iriResolver,
            targetToken: unionToken,
        });
        return UnionPattern.createFrom(newContainer, {});
    };
}
export const GroupPattern = {
    createFrom(container, object) {
        return NotTriplePattern.createFrom(container, Object.assign(object, {
            union: getUnionFn(container),
        }));
    }
};
