// sparqler/.esm2015/patterns/notTriplePatterns/SingleValuesPattern.js
import { cloneElement } from "../../data/utils";
import { convertValue } from "../utils";
import { NotTriplePattern } from "./NotTriplePattern";
function getHasFn(container) {
    return value => {
        const values = container.targetToken.values.slice();
        if (!values.length)
            values.push([]);
        values[0] = values[0].concat(convertValue(value));
        const targetToken = cloneElement(container.targetToken, { values });
        const newContainer = cloneElement(container, { targetToken });
        return SingleValuesPatternMore.createFrom(newContainer, {});
    };
}
export const SingleValuesPattern = {
    createFrom(container, object) {
        return NotTriplePattern.createFrom(container, Object.assign(object, {
            has: getHasFn(container),
        }));
    },
};
export const SingleValuesPatternMore = {
    createFrom(container, object) {
        return NotTriplePattern.createFrom(container, Object.assign(object, {
            and: getHasFn(container),
        }));
    },
};
