// sparqler/.esm5/clauses/SelectClause.js
import { Container } from "../data/Container";
import { cloneElement } from "../data/utils";
import { SelectToken } from "../tokens/SelectToken";
import { VariableToken } from "../tokens/VariableToken";
import { FromClause } from "./FromClause";
function getSelectFn(genericFactory, container, modifier) {
    return function () {
        var variables = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            variables[_i] = arguments[_i];
        }
        var queryClause = new SelectToken(modifier);
        if (variables.length)
            queryClause.addVariable.apply(queryClause, variables.map(function (x) { return new VariableToken(x); }));
        var queryToken = cloneElement(container.targetToken, { queryClause: queryClause });
        var newContainer = new Container({
            iriResolver: container.iriResolver,
            targetToken: queryToken,
        });
        return FromClause.createFrom(genericFactory, newContainer, {});
    };
}
export var SelectClause = {
    createFrom: function (genericFactory, container, object) {
        return Object.assign(object, {
            select: getSelectFn(genericFactory, container),
            selectDistinct: getSelectFn(genericFactory, container, "DISTINCT"),
            selectReduced: getSelectFn(genericFactory, container, "REDUCED"),
            selectAll: function () { return getSelectFn(genericFactory, container)(); },
            selectAllDistinct: function () { return getSelectFn(genericFactory, container, "DISTINCT")(); },
            selectAllReduced: function () { return getSelectFn(genericFactory, container, "REDUCED")(); },
        });
    },
};
