// sparqler/.esm5/patterns/clausePatterns/SubSelectPattern.js
import { Container } from "../../data/Container";
import { SubSelectToken } from "../../tokens/SubSelectToken";
import { VariableToken } from "../../tokens/VariableToken";
import { WherePattern } from "./WherePattern";
function getSelectFn(container, modifier) {
    return function () {
        var variables = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            variables[_i] = arguments[_i];
        }
        var targetToken = new SubSelectToken(modifier);
        if (variables.length)
            targetToken.addVariable.apply(targetToken, variables.map(function (x) { return new VariableToken(x); }));
        var newContainer = new Container({
            iriResolver: container.iriResolver,
            targetToken: targetToken
        });
        return WherePattern.createFrom(newContainer, {});
    };
}
export var SubSelectPattern = {
    createFrom: function (container, object) {
        return Object.assign(object, {
            select: getSelectFn(container),
            selectDistinct: getSelectFn(container, "DISTINCT"),
            selectReduced: getSelectFn(container, "REDUCED"),
            selectAll: function () { return getSelectFn(container)(); },
            selectAllDistinct: function () { return getSelectFn(container, "DISTINCT")(); },
            selectAllReduced: function () { return getSelectFn(container, "REDUCED")(); },
        });
    },
};
