// carbonldp/.esm5/Registry/RegisteredPointer.js
import { IllegalArgumentError } from "../Errors/IllegalArgumentError";
import { ModelDecorator } from "../Model/ModelDecorator";
import { Pointer } from "../Pointer/Pointer";
export var RegisteredPointer = {
    PROTOTYPE: {
        get $registry() {
            throw new IllegalArgumentError("Property \"$registry\" is required.");
        },
    },
    isDecorated: function (object) {
        return ModelDecorator
            .hasPropertiesFrom(RegisteredPointer.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (RegisteredPointer.isDecorated(object))
            return object;
        var resource = ModelDecorator
            .decorateMultiple(object, Pointer);
        return ModelDecorator
            .definePropertiesFrom(RegisteredPointer.PROTOTYPE, resource);
    },
    create: function (data) {
        var copy = Object.assign({}, data);
        return RegisteredPointer.createFrom(copy);
    },
    createFrom: function (object) {
        return RegisteredPointer.decorate(object);
    },
    is: function (value) {
        return Pointer.is(value)
            && RegisteredPointer.isDecorated(value);
    },
};
