// carbonldp/.esm5/Document/Traits/SPARQLDocumentTrait.js
import { ModelDecorator } from "../../Model/ModelDecorator";
import { URI } from "../../RDF/URI";
import { ResolvablePointer } from "../../Repository/ResolvablePointer";
import { isObject } from "../../Utils";
import { TransientDocument } from "../TransientDocument";
function __parseParams(resource, uriOrQuery, queryOrOptions, options) {
    var uri = resource.$id;
    var query = uriOrQuery;
    if (isObject(queryOrOptions)) {
        options = queryOrOptions;
    }
    else if (queryOrOptions !== void 0) {
        query = queryOrOptions;
        uri = URI.resolve(resource.$id, uriOrQuery);
    }
    return { uri: uri, query: query, options: options };
}
export var SPARQLDocumentTrait = {
    PROTOTYPE: {
        $executeASKQuery: function (uriOrQuery, queryOrOptions, requestOptions) {
            var _a = __parseParams(this, uriOrQuery, queryOrOptions, requestOptions), uri = _a.uri, query = _a.query, options = _a.options;
            return this.$repository.executeASKQuery(uri, query, options);
        },
        $executeSELECTQuery: function (uriOrQuery, queryOrOptions, requestOptions) {
            var _a = __parseParams(this, uriOrQuery, queryOrOptions, requestOptions), uri = _a.uri, query = _a.query, options = _a.options;
            return this.$repository.executeSELECTQuery(uri, query, options);
        },
        $executeUPDATE: function (uriOrQuery, updateOrOptions, requestOptions) {
            var _a = __parseParams(this, uriOrQuery, updateOrOptions, requestOptions), uri = _a.uri, query = _a.query, options = _a.options;
            return this.$repository.executeUPDATE(uri, query, options);
        },
        $sparql: function (uri) {
            var $uri = uri ? URI.resolve(this.$id, uri) : this.$id;
            return this.$repository.sparql($uri);
        },
    },
    isDecorated: function (object) {
        return isObject(object)
            && ModelDecorator
                .hasPropertiesFrom(SPARQLDocumentTrait.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (SPARQLDocumentTrait.isDecorated(object))
            return object;
        var target = ModelDecorator
            .decorateMultiple(object, TransientDocument, ResolvablePointer);
        return ModelDecorator
            .definePropertiesFrom(SPARQLDocumentTrait.PROTOTYPE, target);
    },
};
