// carbonldp/.esm5/Repository/ResolvablePointer.js
import { _parseResourceParams, _parseURIParams } from "../DocumentsRepository/Utils";
import { IllegalArgumentError } from "../Errors/IllegalArgumentError";
import { ModelDecorator } from "../Model/ModelDecorator";
import { Pointer } from "../Pointer/Pointer";
import { isFunction, isObject, ObjectUtils } from "../Utils";
function __internalRevert(target, source) {
    if (!isObject(target) || !isObject(source))
        return;
    new Set(Object.keys(target).concat(Object.keys(source))).forEach(function (key) {
        var sourceValue = Array.isArray(source[key]) ? source[key].slice() : source[key];
        if (sourceValue === null || sourceValue === void 0) {
            delete target[key];
            return;
        }
        if (isFunction(sourceValue))
            return;
        target[key] = sourceValue;
    });
}
export var ResolvablePointer = {
    PROTOTYPE: {
        get $repository() {
            throw new IllegalArgumentError("Property \"$repository\" is required.");
        },
        $eTag: void 0,
        $_resolved: false,
        $isResolved: function () {
            return this.$_resolved;
        },
        $_snapshot: {},
        $_syncSnapshot: function () {
            var clone = ObjectUtils.clone(this, { arrays: true });
            if (this.types)
                clone.types = this.types.slice();
            this.$_snapshot = clone;
        },
        $isDirty: function () {
            return !ObjectUtils
                .areEqual(this, this.$_snapshot, { arrays: true });
        },
        $revert: function () {
            __internalRevert(this, this.$_snapshot);
            if (!this.types)
                this.types = [];
        },
        $get: function (uri) {
            var _a, _b;
            var _c = _parseURIParams(this, uri, arguments), _uri = _c._uri, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$get.apply(_a, [_uri].concat(_args)) : (_b = this.$repository).get.apply(_b, [_uri].concat(_args));
        },
        $resolve: function (resource) {
            var _a, _b;
            var _c = _parseResourceParams(this, resource, arguments), _resource = _c._resource, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$resolve.apply(_a, [_resource].concat(_args)) : (_b = this.$repository).resolve.apply(_b, [_resource].concat(_args));
        },
        $exists: function (uri) {
            var _a, _b;
            var _c = _parseURIParams(this, uri, arguments), _uri = _c._uri, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$exists.apply(_a, [_uri].concat(_args)) : (_b = this.$repository).exists.apply(_b, [_uri].concat(_args));
        },
        $refresh: function (resource) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var _a, _b;
            var _c = _parseResourceParams(this, resource, arguments), _resource = _c._resource, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$refresh.apply(_a, [_resource].concat(_args)) : (_b = this.$repository).refresh.apply(_b, [_resource].concat(_args));
        },
        $save: function (resource) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var _a, _b;
            var _c = _parseResourceParams(this, resource, arguments), _resource = _c._resource, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$save.apply(_a, [_resource].concat(_args)) : (_b = this.$repository).save.apply(_b, [_resource].concat(_args));
        },
        $saveAndRefresh: function (resource) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var _a, _b;
            var _c = _parseResourceParams(this, resource, arguments), _resource = _c._resource, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$saveAndRefresh.apply(_a, [_resource].concat(_args)) : (_b = this.$repository).saveAndRefresh.apply(_b, [_resource].concat(_args));
        },
        $delete: function (uri) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var _a, _b;
            var _c = _parseURIParams(this, uri, arguments), _uri = _c._uri, _args = _c._args;
            return "$id" in this.$repository ? (_a = this.$repository).$delete.apply(_a, [_uri].concat(_args)) : (_b = this.$repository).delete.apply(_b, [_uri].concat(_args));
        },
    },
    isDecorated: function (object) {
        return ModelDecorator
            .hasPropertiesFrom(ResolvablePointer.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (ResolvablePointer.isDecorated(object))
            return object;
        var resource = ModelDecorator
            .decorateMultiple(object, Pointer);
        return ModelDecorator
            .definePropertiesFrom(ResolvablePointer.PROTOTYPE, resource);
    },
    is: function (value) {
        return Pointer.is(value)
            && ResolvablePointer.isDecorated(value);
    },
};
