// carbonldp/QueryDocuments/QueryRootProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var DigestedObjectSchemaProperty_1 = require("../ObjectSchema/DigestedObjectSchemaProperty");
var QueryProperty_1 = require("./QueryProperty");
var QueryPropertyType_1 = require("./QueryPropertyType");
var QueryRootProperty = (function (_super) {
    tslib_1.__extends(QueryRootProperty, _super);
    function QueryRootProperty(data) {
        return _super.call(this, {
            queryContainer: data.queryContainer,
            name: "document",
            definition: new DigestedObjectSchemaProperty_1.DigestedObjectSchemaProperty(),
            optional: false,
            propertyType: QueryPropertyType_1.QueryPropertyType.PARTIAL,
            values: data.values,
        }) || this;
    }
    QueryRootProperty.prototype.__createSelfPattern = function () {
        return;
    };
    return QueryRootProperty;
}(QueryProperty_1.QueryProperty));
exports.QueryRootProperty = QueryRootProperty;
