// carbonldp/QueryDocuments/QueryableProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IllegalArgumentError_1 = require("../Errors/IllegalArgumentError");
var DigestedObjectSchema_1 = require("../ObjectSchema/DigestedObjectSchema");
var Utils_1 = require("./Utils");
var QueryableProperty = (function () {
    function QueryableProperty(data) {
        this.definition = data.definition;
        this.pathBuilderFn = data.pathBuilderFn;
        this.propertyType = data.propertyType;
        this.optional = data.optional;
        this.subProperties = new Map();
        this.values = data.values
            ? data.values
            : [];
    }
    QueryableProperty.prototype.setType = function (type) {
        this.propertyType = Utils_1._getBestType(this.propertyType, type);
    };
    QueryableProperty.prototype.setProperty = function (propertyName, property) {
        this.subProperties.set(propertyName, property);
    };
    QueryableProperty.prototype.getProperty = function (propertyName, data) {
        if (!this.subProperties.has(propertyName)) {
            if (!data)
                throw new Error("Property \"" + propertyName + "\" doesn't exists.");
            var property = new QueryableProperty(data);
            this.subProperties.set(propertyName, property);
            return property;
        }
        else {
            var property = this.subProperties.get(propertyName);
            if (data)
                property
                    .mergeData(propertyName, data);
            return property;
        }
    };
    QueryableProperty.prototype.mergeData = function (propertyName, data) {
        if (this === data)
            return;
        this.setType(data.propertyType);
        this.__mergeDefinition(propertyName, data.definition);
    };
    QueryableProperty.prototype.__mergeDefinition = function (propertyName, newDefinition) {
        for (var key in newDefinition) {
            var oldValue = this.definition[key];
            var newValue = newDefinition[key];
            if (oldValue === null)
                this.definition[key] = newValue;
            if (newValue !== oldValue) {
                throw new IllegalArgumentError_1.IllegalArgumentError("Property \"" + propertyName + "\" has different \"" + key + "\": \"" + oldValue + "\", \"" + newValue + "\".");
            }
        }
    };
    QueryableProperty.prototype.getSchema = function () {
        var schema = new DigestedObjectSchema_1.DigestedObjectSchema();
        this.subProperties.forEach(function (property, propertyName) {
            schema.properties.set(propertyName, property.definition);
        });
        return schema;
    };
    return QueryableProperty;
}());
exports.QueryableProperty = QueryableProperty;
