// sparqler/.esm2015/tokens/SubjectToken.js
import { addSpaces, getIndentation, getSeparator, INDENTATION_SPACES } from "./printing";
export class SubjectToken {
    constructor(subject) {
        this.token = "subject";
        this.subject = subject;
        this.properties = [];
    }
    addProperty(property) {
        this.properties.push(property);
        return this;
    }
    toString(spaces) {
        let query = this.subject.toString(spaces);
        let separator = !this.properties.length ? ""
            : (this.subject.token === "collection" || this.subject.token === "blankNodeProperty")
                && query.includes("\n") ? "\n"
                : " ";
        const subSpaces = separator === " " ?
            addSpaces(spaces, query.length + 1) :
            addSpaces(spaces, INDENTATION_SPACES);
        const subIndent = getIndentation(subSpaces);
        const properties = this.properties
            .map(property => property.toString(subSpaces))
            .join(";" + getSeparator(spaces) + subIndent);
        if (separator === "\n")
            separator += subIndent;
        return query + separator + properties;
    }
}
