// sparqler/.esm2015/patterns/triplePatterns/TriplePatternsBuilder.js
import { Container } from "../../data/Container";
import { Factory } from "../../data/Factory";
import { BlankNodePropertyToken } from "../../tokens/BlankNodePropertyToken";
import { BlankNodeToken } from "../../tokens/BlankNodeToken";
import { CollectionToken } from "../../tokens/CollectionToken";
import { LiteralToken } from "../../tokens/LiteralToken";
import { RDFLiteralToken } from "../../tokens/RDFLiteralToken";
import { SubjectToken } from "../../tokens/SubjectToken";
import { VariableToken } from "../../tokens/VariableToken";
import { Pattern } from "../Pattern";
import { convertValue } from "../utils";
import { BlankNodeBuilder } from "./BlankNodeBuilder";
import { RDFLiteral } from "./RDFLiteral";
import { TripleSubject } from "./TripleSubject";
function _getPatternContainer(container, token) {
    return new Container({
        iriResolver: container.iriResolver,
        targetToken: new SubjectToken(token),
    });
}
function _getTripleSubject(container, token) {
    const patternContainer = _getPatternContainer(container, token);
    return TripleSubject.createFrom(patternContainer, {});
}
function _getNodeSubject(container, token) {
    const patternContainer = _getPatternContainer(container, token);
    return Factory.createFrom(TripleSubject.createFrom, Pattern.createFrom)(patternContainer, {});
}
function getResourceFn(container) {
    return iri => {
        const token = container.iriResolver.resolve(iri);
        return _getTripleSubject(container, token);
    };
}
function getVarFn(container) {
    return name => {
        const token = new VariableToken(name);
        return _getTripleSubject(container, token);
    };
}
function getLiteralFn(container) {
    return (value) => {
        if (typeof value !== "string") {
            const token = new LiteralToken(value);
            return _getTripleSubject(container, token);
        }
        const token = new RDFLiteralToken(value);
        const patternContainer = _getPatternContainer(container, token);
        return RDFLiteral.createFrom(patternContainer, {});
    };
}
function getCollectionFn(container) {
    return (...values) => {
        const token = new CollectionToken()
            .addObject(...values.map(convertValue));
        return _getNodeSubject(container, token);
    };
}
function _getBlankNode(container, label) {
    if (label && !label.startsWith("_:"))
        label = "_:" + label;
    const token = new BlankNodeToken(label);
    return _getTripleSubject(container, token);
}
function _getBlankNodeProperty(container, builderFn) {
    const token = new BlankNodePropertyToken();
    const builderContainer = new Container({
        iriResolver: container.iriResolver,
        targetToken: token,
    });
    const builder = BlankNodeBuilder.createFrom(builderContainer, {});
    builderFn(builder);
    if (token.properties.length < 1)
        throw new Error("At least one property must be specified by the self builder.");
    return _getNodeSubject(container, token);
}
function getBlankNodeFn(container) {
    return (labelOrBuilderFn) => {
        if (typeof labelOrBuilderFn === "function")
            return _getBlankNodeProperty(container, labelOrBuilderFn);
        return _getBlankNode(container, labelOrBuilderFn);
    };
}
export const TriplePatternsBuilder = {
    createFrom(container, object) {
        return Object.assign(object, {
            resource: getResourceFn(container),
            var: getVarFn(container),
            literal: getLiteralFn(container),
            collection: getCollectionFn(container),
            blankNode: getBlankNodeFn(container),
        });
    },
};
