// sparqler/.esm5/patterns/notTriplePatterns/NotTriplePatternsBuilder.js
import { Container } from "../../data/Container";
import { BindToken } from "../../tokens/BindToken";
import { FilterToken } from "../../tokens/FilterToken";
import { GraphToken } from "../../tokens/GraphToken";
import { GroupPatternToken } from "../../tokens/GroupPatternToken";
import { MinusPatternToken } from "../../tokens/MinusPatternToken";
import { OptionalToken } from "../../tokens/OptionalToken";
import { ServicePatternToken } from "../../tokens/ServicePatternToken";
import { UnionPatternToken } from "../../tokens/UnionPatternToken";
import { ValuesToken } from "../../tokens/ValuesToken";
import { VariableToken } from "../../tokens/VariableToken";
import { GroupPattern } from "./GroupPattern";
import { MultipleValuesPattern } from "./MultipleValuesPattern";
import { NotTriplePattern } from "./NotTriplePattern";
import { SingleValuesPattern } from "./SingleValuesPattern";
import { UnionPattern } from "./UnionPattern";
function _getPatternContainer(container, targetToken) {
    return new Container({
        iriResolver: container.iriResolver,
        targetToken: targetToken,
    });
}
function _getPattern(container, token) {
    var patternContainer = _getPatternContainer(container, token);
    return NotTriplePattern.createFrom(patternContainer, {});
}
function getGraphFn(container) {
    return function (iriOrVariable, patterns) {
        var varOrIRI = typeof iriOrVariable === "string" ?
            container.iriResolver.resolve(iriOrVariable) :
            iriOrVariable.getSubject();
        var token = new GraphToken(varOrIRI);
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        token.addPattern.apply(token, patterns.map(function (x) { return x.getPattern(); }));
        return _getPattern(container, token);
    };
}
function getGroupFn(container) {
    return function (patterns) {
        var _a;
        var token = new GroupPatternToken();
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        (_a = token.patterns).push.apply(_a, patterns.map(function (x) { return x.getPattern(); }));
        var patternContainer = _getPatternContainer(container, token);
        return GroupPattern.createFrom(patternContainer, {});
    };
}
function getUnionFn(container) {
    return function (patterns) {
        var token = new UnionPatternToken();
        var patternContainer = _getPatternContainer(container, token);
        var unionPattern = UnionPattern
            .createFrom(patternContainer, {});
        return unionPattern.and(patterns);
    };
}
function getOptionalFn(container) {
    return function (patterns) {
        var token = new OptionalToken();
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        token.addPattern.apply(token, patterns.map(function (x) { return x.getPattern(); }));
        return _getPattern(container, token);
    };
}
function getMinusFn(container) {
    return function (patterns) {
        var _a;
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        var token = new MinusPatternToken();
        (_a = token.groupPattern.patterns).push.apply(_a, patterns.map(function (x) { return x.getPattern(); }));
        return _getPattern(container, token);
    };
}
function getServiceFn(container, modifier) {
    return function (resource, patterns) {
        var _a;
        var varOrIRI = typeof resource === "string" ?
            container.iriResolver.resolve(resource) :
            resource.getSubject();
        var token = new ServicePatternToken(varOrIRI, modifier);
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        (_a = token.groupPattern.patterns).push.apply(_a, patterns.map(function (x) { return x.getPattern(); }));
        return _getPattern(container, token);
    };
}
function getFilterFn(container) {
    return function (rawConstraint) {
        var token = new FilterToken(rawConstraint);
        return _getPattern(container, token);
    };
}
function getBindFn(container) {
    return function (rawExpression, variable) {
        var parsedVar = typeof variable === "string" ?
            new VariableToken(variable) :
            variable.getSubject();
        var token = new BindToken(rawExpression, parsedVar);
        return _getPattern(container, token);
    };
}
function getValuesFn(container) {
    return function () {
        var variables = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            variables[_i] = arguments[_i];
        }
        var _a;
        var token = new ValuesToken();
        (_a = token.variables).push.apply(_a, variables.map(function (x) { return x.getSubject(); }));
        var patternContainer = _getPatternContainer(container, token);
        if (variables.length === 1)
            return SingleValuesPattern
                .createFrom(patternContainer, {});
        return MultipleValuesPattern
            .createFrom(patternContainer, {});
    };
}
export var NotTriplePatternsBuilder = {
    createFrom: function (container, object) {
        return Object.assign(object, {
            undefined: "UNDEF",
            graph: getGraphFn(container),
            group: getGroupFn(container),
            union: getUnionFn(container),
            optional: getOptionalFn(container),
            minus: getMinusFn(container),
            service: getServiceFn(container),
            serviceSilent: getServiceFn(container, "SILENT"),
            filter: getFilterFn(container),
            bind: getBindFn(container),
            values: getValuesFn(container),
        });
    },
};
