// sparqler/.esm5/patterns/notTriplePatterns/UnionPattern.js
import { Container } from "../../data/Container";
import { cloneElement } from "../../data/utils";
import { GroupPatternToken } from "../../tokens/GroupPatternToken";
import { NotTriplePattern } from "./NotTriplePattern";
function getAndFn(container) {
    return function (patterns) {
        var _a;
        patterns = Array.isArray(patterns) ? patterns : [patterns];
        var newGroupToken = new GroupPatternToken();
        (_a = newGroupToken.patterns).push.apply(_a, patterns.map(function (x) { return x.getPattern(); }));
        var groupPatterns = container.targetToken.groupPatterns.concat(newGroupToken);
        var unionToken = cloneElement(container.targetToken, { groupPatterns: groupPatterns });
        var newContainer = new Container({
            iriResolver: container.iriResolver,
            targetToken: unionToken,
        });
        return UnionPattern.createFrom(newContainer, {});
    };
}
export var UnionPattern = {
    createFrom: function (container, object) {
        return NotTriplePattern.createFrom(container, Object.assign(object, {
            and: getAndFn(container),
        }));
    },
};
