// sparqler/.esm2015/clauses/FromClause.js
import { IRIResolver } from "../data/IRIResolver";
import { cloneElement } from "../data/utils";
import { FromToken } from "../tokens/FromToken";
import { WhereClause } from "./WhereClause";
function getFromFn(genericFactory, container, named) {
    return (iri) => {
        const iriResolver = new IRIResolver(container.iriResolver);
        const datasets = container.targetToken.queryClause.datasets
            .concat(new FromToken(iriResolver.resolve(iri), named));
        const queryClause = cloneElement(container.targetToken.queryClause, { datasets });
        const queryToken = cloneElement(container.targetToken, { queryClause });
        const newContainer = cloneElement(container, {
            iriResolver,
            targetToken: queryToken,
        });
        return FromClause.createFrom(genericFactory, newContainer, {});
    };
}
export const FromClause = {
    createFrom(genericFactory, container, object) {
        return WhereClause.createFrom(genericFactory, container, Object.assign(object, {
            from: getFromFn(genericFactory, container),
            fromNamed: getFromFn(genericFactory, container, true),
        }));
    },
};
