// sparqler/.esm2015/clauses/QueryClause.js
import { Factory } from "../data/Factory";
import { IRIResolver } from "../data/IRIResolver";
import { cloneElement } from "../data/utils";
import { BaseToken } from "../tokens/BaseToken";
import { IRIRefToken } from "../tokens/IRIRefToken";
import { PrefixToken } from "../tokens/PrefixToken";
import { AskClause } from "./AskClause";
import { SelectClause } from "./SelectClause";
function base(iri) {
    const token = new BaseToken(new IRIRefToken(iri));
    const prologues = this.targetToken
        .prologues.concat(token);
    const queryToken = cloneElement(this.targetToken, { prologues });
    const container = cloneElement(this, { targetToken: queryToken });
    return QueryClause.createFrom(container, {});
}
function vocab(iri) {
    const iriResolver = new IRIResolver(this.iriResolver, iri);
    const container = cloneElement(this, { iriResolver });
    return QueryClause.createFrom(container, {});
}
function prefix(name, iri) {
    const iriResolver = new IRIResolver(this.iriResolver);
    const prologues = this.targetToken.prologues.slice();
    if (iriResolver.prefixes.has(name)) {
        const index = prologues
            .findIndex(token => token.token === "prefix" && token.namespace === name);
        if (index !== -1)
            prologues.splice(index, 1);
    }
    prologues.push(new PrefixToken(name, new IRIRefToken(iri)));
    iriResolver.prefixes.set(name, false);
    const queryToken = cloneElement(this.targetToken, { prologues });
    const container = cloneElement(this, {
        iriResolver,
        targetToken: queryToken,
    });
    return QueryClause.createFrom(container, {});
}
export const QueryClause = {
    createFrom(container, object) {
        const selectFactory = SelectClause
            .createFrom.bind(null, container.selectFinishClauseFactory);
        const askFactory = AskClause
            .createFrom.bind(null, container.askFinishClauseFactory);
        return Factory.createFrom(selectFactory, askFactory)(container, Object.assign(object, {
            base: base.bind(container),
            vocab: vocab.bind(container),
            prefix: prefix.bind(container),
        }));
    },
};
