// sparqler/.esm5/tokens/SubjectToken.js
import { addSpaces, getIndentation, getSeparator, INDENTATION_SPACES } from "./printing";
var SubjectToken = (function () {
    function SubjectToken(subject) {
        this.token = "subject";
        this.subject = subject;
        this.properties = [];
    }
    SubjectToken.prototype.addProperty = function (property) {
        this.properties.push(property);
        return this;
    };
    SubjectToken.prototype.toString = function (spaces) {
        var query = this.subject.toString(spaces);
        var separator = !this.properties.length ? ""
            : (this.subject.token === "collection" || this.subject.token === "blankNodeProperty")
                && query.includes("\n") ? "\n"
                : " ";
        var subSpaces = separator === " " ?
            addSpaces(spaces, query.length + 1) :
            addSpaces(spaces, INDENTATION_SPACES);
        var subIndent = getIndentation(subSpaces);
        var properties = this.properties
            .map(function (property) { return property.toString(subSpaces); })
            .join(";" + getSeparator(spaces) + subIndent);
        if (separator === "\n")
            separator += subIndent;
        return query + separator + properties;
    };
    return SubjectToken;
}());
export { SubjectToken };
