// carbonldp/.esm2015/LDP/DirectContainer/TransientDirectContainer.js
import { TransientDocument } from "../../Document/TransientDocument";
import { IllegalArgumentError } from "../../Errors/IllegalArgumentError";
import { LDP } from "../../Vocabularies/LDP";
export const TransientDirectContainer = {
    TYPE: LDP.DirectContainer,
    is(value) {
        return TransientDocument.is(value)
            && value.$hasType(TransientDirectContainer.TYPE)
            && value.hasOwnProperty("hasMemberRelation");
    },
    create(data) {
        const copy = Object.assign({}, data);
        return TransientDirectContainer.createFrom(copy);
    },
    createFrom(object) {
        if (TransientDirectContainer.is(object))
            throw new IllegalArgumentError("The base object is already a DirectContainer.");
        if (!object.hasMemberRelation)
            throw new IllegalArgumentError("The property hasMemberRelation is required.");
        const container = TransientDocument.is(object) ?
            object : TransientDocument.createFrom(object);
        container.$addType(TransientDirectContainer.TYPE);
        return container;
    },
};
