// carbonldp/.esm5/QueryDocuments/QueryContainer.js
import * as tslib_1 from "tslib";
import { IRIResolver } from "sparqler/data";
import { isPrefixed } from "sparqler/iri";
import { DeniableFluentPath, FluentPath, FluentPathContainer } from "sparqler/patterns";
import { IRIRefToken, PrefixToken } from "sparqler/tokens";
import { IllegalArgumentError } from "../Errors/IllegalArgumentError";
import { ObjectSchemaDigester } from "../ObjectSchema/ObjectSchemaDigester";
import { QueryContainerProperty } from "./QueryContainerProperty";
import { QueryRootProperty } from "./QueryRootProperty";
import { QueryVariable } from "./QueryVariable";
var QueryContainer = (function (_super) {
    tslib_1.__extends(QueryContainer, _super);
    function QueryContainer(context, propertyData) {
        var _this = this;
        var schema = context.getObjectSchema();
        _this = _super.call(this, {
            iriResolver: __createIRIResolver(schema),
            targetToken: void 0,
            fluentPathFactory: FluentPath.createFrom,
            deniableFluentPathFactory: DeniableFluentPath.createFrom,
        }) || this;
        _this.context = context;
        _this._generalSchema = schema;
        _this._prefixesTuples = Array.from(schema.prefixes);
        _this._variablesCounter = 0;
        _this._variablesMap = new Map();
        if ("uris" in propertyData) {
            var values = propertyData.uris.map(_this.compactIRI, _this);
            _this._queryProperty = new QueryRootProperty({
                queryContainer: _this,
                values: values,
            });
        }
        else {
            var iri = _this.compactIRI(propertyData.uri);
            _this._queryProperty = new QueryContainerProperty({
                queryContainer: _this,
                containerIRI: iri,
                containerPropertyType: propertyData.containerPropertyType,
            });
        }
        return _this;
    }
    QueryContainer.prototype.getVariable = function (name) {
        if (this._variablesMap.has(name))
            return this._variablesMap.get(name);
        var variable = new QueryVariable(name, this._variablesCounter++);
        this._variablesMap.set(name, variable);
        return variable;
    };
    QueryContainer.prototype.compactIRI = function (iri) {
        var compactedIRI = this.__getCompactedIRI(iri);
        return this.iriResolver.resolve(compactedIRI);
    };
    QueryContainer.prototype.__getCompactedIRI = function (iri) {
        if (isPrefixed(iri))
            return iri;
        var prefix = this._prefixesTuples
            .find(function (_a) {
            var x = _a[1];
            return iri.startsWith(x);
        });
        if (!prefix)
            return iri;
        var namespace = prefix[0], prefixIRI = prefix[1];
        return namespace + ":" + iri.substr(prefixIRI.length);
    };
    QueryContainer.prototype.getPrologues = function () {
        return this._prefixesTuples
            .filter(this.__isUsedPrefix, this)
            .map(__createPrefixToken);
    };
    QueryContainer.prototype.__isUsedPrefix = function (_a) {
        var namespace = _a[0];
        return !!this.iriResolver.prefixes.get(namespace);
    };
    QueryContainer.prototype.digestProperty = function (name, definition) {
        return ObjectSchemaDigester
            .digestProperty(name, definition, this._generalSchema);
    };
    QueryContainer.prototype.getGeneralSchema = function () {
        return ObjectSchemaDigester
            .combineDigestedObjectSchemas([this._generalSchema]);
    };
    QueryContainer.prototype.serializeLiteral = function (type, value) {
        if (!this.context.jsonldConverter.literalSerializers.has(type))
            throw new IllegalArgumentError("Type \"" + type + "\" hasn't a defined serializer.");
        return this.context.jsonldConverter
            .literalSerializers
            .get(type)
            .serialize(value);
    };
    return QueryContainer;
}(FluentPathContainer));
export { QueryContainer };
function __createIRIResolver(schema) {
    var iriResolver = new IRIResolver(void 0, schema.vocab);
    Array.from(schema.prefixes.keys())
        .forEach(function (key) { return iriResolver.prefixes.set(key, false); });
    return iriResolver;
}
function __createPrefixToken(_a) {
    var namespace = _a[0], iri = _a[1];
    return new PrefixToken(namespace, new IRIRefToken(iri));
}
