// carbonldp/GeneralRepository/GeneralRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var IllegalArgumentError_1 = require("../Errors/IllegalArgumentError");
var ModelDecorator_1 = require("../Model/ModelDecorator");
var Repository_1 = require("../Repository/Repository");
exports.GeneralRepository = {
    PROTOTYPE: {
        get context() {
            throw new IllegalArgumentError_1.IllegalArgumentError("Property \"context\" is required.");
        },
    },
    isDecorated: function (object) {
        return ModelDecorator_1.ModelDecorator
            .hasPropertiesFrom(exports.GeneralRepository.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (exports.GeneralRepository.isDecorated(object))
            return object;
        var target = ModelDecorator_1.ModelDecorator
            .decorateMultiple(object, Repository_1.Repository);
        return ModelDecorator_1.ModelDecorator
            .definePropertiesFrom(exports.GeneralRepository.PROTOTYPE, target);
    },
    create: function (data) {
        return exports.GeneralRepository.createFrom(tslib_1.__assign({}, data));
    },
    createFrom: function (object) {
        return exports.GeneralRepository.decorate(object);
    },
};
