// carbonldp/HTTP/Response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BadResponseError_1 = require("./Errors/ServerErrors/BadResponseError");
var Header_1 = require("./Header");
var Response = (function () {
    function Response(request, data, response) {
        this.request = request;
        if (typeof XMLHttpRequest !== "undefined" && request instanceof XMLHttpRequest) {
            this.status = request.status;
            this.data = request.responseText;
            this.headers = Header_1.Header.parseHeaders(request.getAllResponseHeaders());
        }
        else {
            this.data = data || "";
            this.status = response && response.statusCode || 0;
            this.headers = new Map();
            if (!response)
                return;
            for (var name_1 in response.headers) {
                var header = new Header_1.Header(response.headers[name_1]);
                this.headers.set(name_1.toLowerCase(), header);
            }
        }
    }
    Response.prototype.getHeader = function (name) {
        name = name.toLowerCase();
        return this.headers.get(name) || null;
    };
    Response.prototype.getETag = function () {
        var eTagHeader = this.getHeader("ETag");
        if (!eTagHeader || !eTagHeader.values.length)
            throw new BadResponseError_1.BadResponseError("The response doesn't contain an ETag", this);
        return eTagHeader.values[0];
    };
    return Response;
}());
exports.Response = Response;
