// carbonldp/.esm5/DocumentsRegistry/DocumentsRegistry.js
import * as tslib_1 from "tslib";
import { Document } from "../Document/Document";
import { IllegalArgumentError } from "../Errors/IllegalArgumentError";
import { GeneralRegistry } from "../GeneralRegistry/GeneralRegistry";
import { ModelDecorator } from "../Model/ModelDecorator";
import { URI } from "../RDF/URI";
export var DocumentsRegistry = {
    PROTOTYPE: {
        register: function (id) {
            return this.getPointer(id, true);
        },
        _getLocalID: function (id) {
            if (URI.hasFragment(id))
                throw new IllegalArgumentError("\"" + id + "\" is out of scope.");
            return GeneralRegistry.PROTOTYPE._getLocalID.call(this, id);
        },
    },
    isDecorated: function (object) {
        return ModelDecorator
            .hasPropertiesFrom(DocumentsRegistry.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (DocumentsRegistry.isDecorated(object))
            return object;
        var base = Object.assign(object, {
            __modelDecorator: Document,
        });
        var target = ModelDecorator
            .decorateMultiple(base, GeneralRegistry);
        return ModelDecorator
            .definePropertiesFrom(DocumentsRegistry.PROTOTYPE, target);
    },
    create: function (data) {
        return DocumentsRegistry.createFrom(tslib_1.__assign({}, data));
    },
    createFrom: function (object) {
        var registry = DocumentsRegistry.decorate(object);
        return GeneralRegistry.createFrom(registry);
    },
};
