// carbonldp/.esm5/ObjectSchema/DigestedObjectSchema.js
import { URI } from "../RDF/URI";
import { ObjectSchemaUtils } from "./ObjectSchemaUtils";
var DigestedObjectSchema = (function () {
    function DigestedObjectSchema() {
        this.base = "";
        this.vocab = undefined;
        this.language = null;
        this.prefixes = new Map();
        this.properties = new Map();
    }
    DigestedObjectSchema.prototype.resolveURI = function (uri, relativeTo) {
        if (relativeTo === void 0) { relativeTo = {}; }
        if (uri === null || URI.isAbsolute(uri) || URI.isBNodeID(uri))
            return uri;
        var _a = uri.split(":"), prefix = _a[0], _b = _a[1], localName = _b === void 0 ? "" : _b;
        var definedReference = this.prefixes.has(prefix) ?
            this.prefixes.get(prefix) : this.properties.has(prefix) ?
            this.properties.get(prefix).uri
            : null;
        if (definedReference !== null && definedReference !== prefix) {
            return this.resolveURI(definedReference + localName, { vocab: true });
        }
        if (localName)
            return uri;
        if (relativeTo.vocab && this.vocab)
            return this.vocab + uri;
        if (relativeTo.base)
            return URI.resolve(this.base, uri);
        return uri;
    };
    DigestedObjectSchema.prototype.getProperty = function (name) {
        if (!this.properties.has(name))
            return void 0;
        return ObjectSchemaUtils._resolveProperty(this, this.properties.get(name));
    };
    return DigestedObjectSchema;
}());
export { DigestedObjectSchema };
