// carbonldp/.esm5/ObjectSchema/ObjectSchemaResolver.js
import { ModelDecorator } from "../Model/ModelDecorator";
import { RDFNode } from "../RDF/Node";
import { URI } from "../RDF/URI";
import { C } from "../Vocabularies/C";
import { DigestedObjectSchema } from "./DigestedObjectSchema";
import { ObjectSchemaDigester } from "./ObjectSchemaDigester";
function __getSchemaForNode($context, node) {
    var types = RDFNode.getTypes(node);
    return __getSchema($context, types, node["@id"]);
}
function __getSchemaForResource($context, resource) {
    var types = resource.types || [];
    return __getSchema($context, types, resource.$id);
}
function __getSchema($context, objectTypes, objectID) {
    if (!$context)
        return new DigestedObjectSchema();
    if (objectID !== void 0 && !URI.hasFragment(objectID) && !URI.isBNodeID(objectID) && objectTypes.indexOf(C.Document) === -1)
        objectTypes = objectTypes.concat(C.Document);
    var objectSchemas = objectTypes
        .filter(function (type) { return $context.hasObjectSchema(type); })
        .map(function (type) { return $context.getObjectSchema(type); });
    return ObjectSchemaDigester
        ._combineSchemas([
        $context.getObjectSchema()
    ].concat(objectSchemas));
}
export var ObjectSchemaResolver = {
    PROTOTYPE: {
        context: undefined,
        getGeneralSchema: function () {
            if (!this.context)
                return new DigestedObjectSchema();
            return this.context.getObjectSchema();
        },
        hasSchemaFor: function (object, path) {
            return !path;
        },
        getSchemaFor: function (object) {
            return "types" in object || "$id" in object ?
                __getSchemaForResource(this.context, object) :
                __getSchemaForNode(this.context, object);
        },
    },
    isDecorated: function (object) {
        return ModelDecorator.hasPropertiesFrom(ObjectSchemaResolver.PROTOTYPE, object);
    },
    decorate: function (object) {
        return ModelDecorator.definePropertiesFrom(ObjectSchemaResolver.PROTOTYPE, object);
    },
};
