// carbonldp/.esm5/RDF/Document.js
import * as Utils from "./../Utils";
import { RDFNode } from "./Node";
import { URI } from "./URI";
export var RDFDocument = {
    is: function (value) {
        return Utils.hasProperty(value, "@graph")
            && Utils.isArray(value["@graph"]);
    },
    create: function (resources, uri) {
        return {
            "@id": uri ? uri : "",
            "@graph": resources,
        };
    },
    getDocuments: function (objects) {
        if (Utils.isArray(objects))
            return objects
                .filter(RDFDocument.is);
        if (RDFDocument.is(objects))
            return [objects];
        return [];
    },
    getFreeNodes: function (objects) {
        if (!Array.isArray(objects))
            return [];
        return objects
            .filter(function (element) { return !RDFDocument.is(element); })
            .filter(RDFNode.is);
    },
    getResources: function (objects) {
        var resources = RDFDocument.getFreeNodes(objects);
        RDFDocument
            .getDocuments(objects)
            .map(function (document) { return document["@graph"]; })
            .forEach(function (nodes) { return resources.push.apply(resources, nodes); });
        return resources;
    },
    getDocumentResources: function (document) {
        return RDFDocument
            .getResources(document)
            .filter(function (node) { return !RDFNode.isFragment(node); });
    },
    getNamedFragmentResources: function (document, documentResource) {
        var uriToMatch = Utils.isObject(documentResource) ?
            RDFNode.getID(documentResource) :
            documentResource;
        return RDFDocument
            .getResources(document)
            .filter(function (node) {
            var id = RDFNode.getID(node);
            if (!URI.hasFragment(id))
                return;
            if (!uriToMatch)
                return true;
            return URI.getDocumentURI(id) === uriToMatch;
        });
    },
    getBNodeResources: function (document) {
        return RDFDocument
            .getResources(document)
            .filter(function (node) {
            var id = RDFNode.getID(node);
            return URI.isBNodeID(id);
        });
    },
    getNodes: function (rdfDocument) {
        var documentNodes = [];
        var fragmentNodes = [];
        for (var _i = 0, _a = rdfDocument["@graph"]; _i < _a.length; _i++) {
            var node = _a[_i];
            (RDFNode.isFragment(node) ? fragmentNodes : documentNodes).push(node);
        }
        return [documentNodes, fragmentNodes];
    },
};
