// carbonldp/.esm5/RDF/Node.js
import * as Utils from "../Utils";
import { XSD } from "../Vocabularies/XSD";
import { RDFList } from "./List";
import { RDFLiteral } from "./Literal";
import { URI } from "./URI";
export var RDFNode = {
    is: function (value) {
        return Utils.hasProperty(value, "@id")
            && Utils.isString(value["@id"]);
    },
    create: function (uri) {
        return {
            "@id": uri,
        };
    },
    getID: function (node) {
        return node["@id"];
    },
    getRelativeID: function (node) {
        var id = RDFNode.getID(node);
        return URI.hasFragment(id) ? URI.getFragment(id) : id;
    },
    areEqual: function (node1, node2) {
        return RDFNode.getID(node1) === RDFNode.getID(node2);
    },
    isFragment: function (node) {
        var id = RDFNode.getID(node);
        return URI.hasFragment(id) || URI.isBNodeID(id);
    },
    hasType: function (node, type) {
        return RDFNode.getTypes(node).indexOf(type) !== -1;
    },
    getTypes: function (node) {
        if (!("@type" in node))
            return [];
        return node["@type"];
    },
    getList: function (propertyValues) {
        if (!Array.isArray(propertyValues))
            return;
        return propertyValues
            .find(RDFList.is);
    },
    getPropertyLiterals: function (propertyValues, literalType) {
        if (!Array.isArray(propertyValues))
            return;
        return propertyValues
            .filter(RDFLiteral.is)
            .filter(function (literal) { return RDFLiteral.hasType(literal, literalType); })
            .map(RDFLiteral.parse);
    },
    getPropertyLanguageMap: function (propertyValues) {
        if (!Array.isArray(propertyValues))
            return;
        var propertyLanguageMap = {};
        for (var _i = 0, propertyValues_1 = propertyValues; _i < propertyValues_1.length; _i++) {
            var propertyValue = propertyValues_1[_i];
            if (!RDFLiteral.is(propertyValue))
                continue;
            if (!RDFLiteral.hasType(propertyValue, XSD.string))
                continue;
            var languageTag = propertyValue["@language"];
            if (!languageTag)
                continue;
            propertyLanguageMap[languageTag] = RDFLiteral.parse(propertyValue);
        }
        return propertyLanguageMap;
    },
};
