// carbonldp/.esm5/RDF/URI.js
import { IllegalArgumentError } from "../Errors/IllegalArgumentError";
import { isString, StringUtils, UUIDUtils } from "../Utils";
export var URI = {
    hasFragment: function (uri) {
        return uri.indexOf("#") !== -1;
    },
    hasQuery: function (uri) {
        return uri.indexOf("?") !== -1;
    },
    hasProtocol: function (uri) {
        return StringUtils.startsWith(uri, "https://") || StringUtils.startsWith(uri, "http://");
    },
    isAbsolute: function (uri) {
        return StringUtils.startsWith(uri, "http://")
            || StringUtils.startsWith(uri, "https://")
            || StringUtils.startsWith(uri, "://");
    },
    isRelative: function (uri) {
        return !URI.isAbsolute(uri);
    },
    isBNodeID: function (uri) {
        return StringUtils.startsWith(uri, "_:");
    },
    generateBNodeID: function () {
        return "_:" + UUIDUtils.generate();
    },
    isPrefixed: function (uri) {
        return !URI.isAbsolute(uri) && !URI.isBNodeID(uri) && StringUtils.contains(uri, ":");
    },
    isFragmentOf: function (fragmentURI, uri) {
        if (!URI.hasFragment(fragmentURI))
            return false;
        var documentURI = URI.getDocumentURI(fragmentURI);
        return documentURI === "" || documentURI === uri;
    },
    isBaseOf: function (baseURI, uri) {
        if (baseURI === uri)
            return true;
        if (baseURI === "")
            return true;
        if (URI.isRelative(uri) && !URI.isPrefixed(uri))
            return true;
        if (uri.startsWith(baseURI)) {
            if (StringUtils.endsWith(baseURI, "/") || StringUtils.endsWith(baseURI, "#"))
                return true;
            var relativeURI = uri.substring(baseURI.length);
            if (StringUtils.startsWith(relativeURI, "/") || StringUtils.startsWith(relativeURI, "#"))
                return true;
        }
        return false;
    },
    getRelativeURI: function (absoluteURI, base) {
        if (!absoluteURI.startsWith(base))
            return absoluteURI;
        return absoluteURI.substring(base.length);
    },
    getDocumentURI: function (uri) {
        var parts = uri.split("#");
        if (parts.length > 2)
            throw new IllegalArgumentError("The URI provided has more than one # sign.");
        return parts[0];
    },
    getFragment: function (uri) {
        var parts = uri.split("#");
        if (parts.length < 2)
            throw new IllegalArgumentError("The URI provided hasn't a # sign.");
        if (parts.length > 2)
            throw new IllegalArgumentError("The URI provided has more than one # sign.");
        return parts[1];
    },
    getSlug: function (uri) {
        var uriParts = uri.split("#");
        if (uriParts.length === 2)
            return URI.getSlug(uriParts[1]);
        if (uriParts.length > 2)
            throw new IllegalArgumentError("Invalid URI: The uri contains two '#' symbols.");
        uri = uriParts[0];
        if (uri === "")
            return uri;
        if (uri === "/")
            return "";
        var parts = uri.split("/");
        if (parts[parts.length - 1] === "") {
            return parts[parts.length - 2];
        }
        else {
            return parts[parts.length - 1];
        }
    },
    getParameters: function (uri) {
        var parameters = new Map();
        if (!URI.hasQuery(uri))
            return parameters;
        uri.replace(/^.*\?/, "").split("&").forEach(function (param) {
            var parts = param
                .replace(/\+/g, " ")
                .split("=");
            var key = parts.shift();
            var val = parts.length > 0 ? parts.join("=") : "";
            if (!parameters.has(key)) {
                parameters.set(key, val);
            }
            else {
                parameters.set(key, new Array().concat(parameters.get(key), val));
            }
        });
        return parameters;
    },
    resolve: function (parentURI, childURI) {
        if (!parentURI || URI.isAbsolute(childURI) || URI.isBNodeID(childURI) || URI.isPrefixed(childURI))
            return childURI;
        var protocol = parentURI.substr(0, parentURI.indexOf("://") + 3);
        var path = parentURI.substr(parentURI.indexOf("://") + 3, parentURI.length - 1);
        if (path.lastIndexOf("/") === -1)
            path += "/";
        if (StringUtils.startsWith(childURI, "?") || StringUtils.startsWith(childURI, "#")) {
            if (URI.hasQuery(path))
                path = path.substr(0, path.indexOf("?"));
            if (URI.hasFragment(path) && (!StringUtils.startsWith(childURI, "?") || StringUtils.endsWith(path, "#")))
                path = URI.getDocumentURI(path);
        }
        else {
            path = path.substr(0, path.lastIndexOf("/") + 1);
            if (!StringUtils.endsWith(path, "?") && !StringUtils.endsWith(path, "#") && !StringUtils.endsWith(path, "/"))
                path += "/";
        }
        if (StringUtils.startsWith(childURI, "/")) {
            childURI = childURI.substr(1, childURI.length);
        }
        return protocol + path + childURI;
    },
    removeProtocol: function (uri) {
        if (!URI.hasProtocol(uri))
            return uri;
        return uri.substring(uri.indexOf("://") + 3);
    },
    prefix: function (uri, prefixOrObjectSchema, prefixURI) {
        if (!isString(prefixOrObjectSchema))
            return prefixWithObjectSchema(uri, prefixOrObjectSchema);
        var prefix = prefixOrObjectSchema;
        if (URI.isPrefixed(uri) || !uri.startsWith(prefixURI))
            return uri;
        return prefix + ":" + uri.substring(prefixURI.length);
    },
};
function prefixWithObjectSchema(uri, objectSchema) {
    var prefixEntries = objectSchema.prefixes.entries();
    while (true) {
        var result = prefixEntries.next();
        if (result.done)
            return uri;
        var _a = result.value, prefix = _a[0], prefixURI = _a[1];
        if (!URI.isAbsolute(prefixURI))
            continue;
        if (!uri.startsWith(prefixURI))
            continue;
        return URI.prefix(uri, prefix, prefixURI);
    }
}
