// carbonldp/.esm5/Pointer/Pointer.js
import { ModelDecorator } from "../Model/ModelDecorator";
import { isObject } from "../Utils";
export var Pointer = {
    PROTOTYPE: {
        get $id() { return ""; },
    },
    isDecorated: function (object) {
        return ModelDecorator
            .hasPropertiesFrom(Pointer.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (Pointer.isDecorated(object))
            return object;
        return ModelDecorator
            .definePropertiesFrom(Pointer.PROTOTYPE, object);
    },
    is: function (value) {
        return isObject(value)
            && Pointer.isDecorated(value);
    },
    create: function (data) {
        var clone = Object.assign({}, data);
        return Pointer.createFrom(clone);
    },
    createFrom: function (object) {
        return Pointer.decorate(object);
    },
    areEqual: function (pointer1, pointer2) {
        return pointer1.$id === pointer2.$id;
    },
    getIDs: function (pointers) {
        return pointers
            .map(function (pointer) { return pointer.$id; });
    },
    getID: function (pointerOrIRI) {
        return isObject(pointerOrIRI) ? pointerOrIRI.$id : pointerOrIRI;
    },
};
