// carbonldp/DocumentsRepository/Traits/QueryableDocumentsRepositoryTrait.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tokens_1 = require("sparqler/tokens");
var IllegalArgumentError_1 = require("../../Errors/IllegalArgumentError");
var IllegalStateError_1 = require("../../Errors/IllegalStateError");
var FreeResources_1 = require("../../FreeResources/FreeResources");
var Request_1 = require("../../HTTP/Request");
var JSONLDParser_1 = require("../../JSONLD/JSONLDParser");
var ModelDecorator_1 = require("../../Model/ModelDecorator");
var Pointer_1 = require("../../Pointer/Pointer");
var QueryContainer_1 = require("../../QueryDocuments/QueryContainer");
var QueryContainerPropertyType_1 = require("../../QueryDocuments/QueryContainerPropertyType");
var QueryDocumentBuilder_1 = require("../../QueryDocuments/QueryDocumentBuilder");
var QueryDocumentsBuilder_1 = require("../../QueryDocuments/QueryDocumentsBuilder");
var QueryMetadata_1 = require("../../QueryDocuments/QueryMetadata");
var QueryPropertyType_1 = require("../../QueryDocuments/QueryPropertyType");
var QueryResultCompacter_1 = require("../../QueryDocuments/QueryResultCompacter");
var Utils_1 = require("../../QueryDocuments/Utils");
var Document_1 = require("../../RDF/Document");
var URI_1 = require("../../RDF/URI");
var SPARQLService_1 = require("../../SPARQL/SPARQLService");
var Utils_2 = require("../../Utils");
var C_1 = require("../../Vocabularies/C");
var Utils_3 = require("../Utils");
var LDPDocumentsRepositoryTrait_1 = require("./LDPDocumentsRepositoryTrait");
function __executeQueryBuilder(queryContainer, queryData) {
    var queryBuilder = "containerPropertyType" in queryContainer._queryProperty
        ? new QueryDocumentsBuilder_1.QueryDocumentsBuilder(queryContainer, queryContainer._queryProperty)
        : new QueryDocumentBuilder_1.QueryDocumentBuilder(queryContainer, queryContainer._queryProperty);
    if (queryData.rootType !== void 0)
        queryContainer._queryProperty
            .setType(queryData.rootType);
    if (queryData.queryBuilderFn && queryData.queryBuilderFn.call(void 0, queryBuilder) !== queryBuilder)
        throw new IllegalArgumentError_1.IllegalArgumentError("The provided query builder was not returned");
}
function __sortQueryDocuments(queryContainer, documents) {
    if (!("order" in queryContainer._queryProperty) || !queryContainer._queryProperty.order)
        return documents;
    var _a = queryContainer._queryProperty.order, path = _a.path, flow = _a.flow;
    var inverter = flow === "DESC" ? -1 : 1;
    return documents.sort(function (a, b) {
        a = Utils_1._getPathProperty(a, path);
        b = Utils_1._getPathProperty(b, path);
        var aValue = Pointer_1.Pointer.is(a) ? a.$id : a;
        var bValue = Pointer_1.Pointer.is(b) ? b.$id : b;
        if (aValue === bValue)
            return 0;
        if (aValue === void 0)
            return -1 * inverter;
        if (bValue === void 0)
            return inverter;
        if (!Utils_1._areDifferentType(a, b)) {
            if (Pointer_1.Pointer.is(a)) {
                var aIsBNode = URI_1.URI.isBNodeID(aValue);
                var bIsBNode = URI_1.URI.isBNodeID(bValue);
                if (aIsBNode && !bIsBNode)
                    return -1 * inverter;
                if (bIsBNode && !aIsBNode)
                    return inverter;
            }
        }
        else {
            if (Pointer_1.Pointer.is(a))
                return -1 * inverter;
            if (Pointer_1.Pointer.is(b))
                return inverter;
            if (Utils_2.isNumber(a))
                return -1 * inverter;
            if (Utils_2.isNumber(b))
                return inverter;
            if (Utils_2.isDate(a))
                return -1 * inverter;
            if (Utils_2.isDate(b))
                return inverter;
            if (Utils_2.isBoolean(a))
                return -1 * inverter;
            if (Utils_2.isBoolean(b))
                return inverter;
            if (Utils_2.isString(a))
                return -1 * inverter;
            if (Utils_2.isString(b))
                return inverter;
        }
        if (aValue < bValue)
            return -1 * inverter;
        if (aValue > bValue)
            return inverter;
        return 0;
    });
}
function __requestQueryDocuments(repository, url, requestOptions, queryContainer) {
    var _a, _b, _c;
    var construct = (_a = (_b = new tokens_1.ConstructToken()
        .addTriple(new tokens_1.SubjectToken(new tokens_1.IRIRefToken("cldp-sdk://metadata-" + Utils_2.UUIDUtils.generate()))
        .addProperty(new tokens_1.PropertyToken("a")
        .addObject(queryContainer.compactIRI(C_1.C.VolatileResource))
        .addObject(queryContainer.compactIRI(C_1.C.QueryMetadata)))
        .addProperty(new tokens_1.PropertyToken(queryContainer.compactIRI(C_1.C.target))
        .addObject(queryContainer._queryProperty.identifier)))).addTriple.apply(_b, queryContainer._queryProperty.getConstructPatterns())).addPattern.apply(_a, queryContainer._queryProperty.getSearchPatterns());
    var query = (_c = new tokens_1.QueryToken(construct)).addPrologues.apply(_c, queryContainer.getPrologues());
    Request_1.RequestUtils.setRetrievalPreferences({ include: [C_1.C.PreferResultsContexts] }, requestOptions);
    return SPARQLService_1.SPARQLService
        .executeRawCONSTRUCTQuery(url, query.toString(), requestOptions)
        .then(function (_a) {
        var strConstruct = _a[0];
        return strConstruct;
    })
        .then(function (jsonldString) {
        return new JSONLDParser_1.JSONLDParser().parse(jsonldString);
    })
        .then(function (rdfNodes) {
        var freeNodes = Document_1.RDFDocument.getFreeNodes(rdfNodes);
        var freeResources = FreeResources_1.FreeResources
            .parseFreeNodes(repository.context.registry, freeNodes);
        var targetDocuments = freeResources
            .getPointers(true)
            .filter(QueryMetadata_1.QueryMetadata.is)
            .map(function (x) { return x.targets; })
            .reduce(function (targets, x) { return targets.concat(x); }, [])
            .map(function (x) { return x.$id; });
        var rdfDocuments = rdfNodes
            .filter(Document_1.RDFDocument.is);
        return new QueryResultCompacter_1.QueryResultCompacter(repository.context.registry, queryContainer)
            .compactDocuments(rdfDocuments, targetDocuments);
    })
        .then(function (documents) { return __sortQueryDocuments(queryContainer, documents); })
        .catch(Utils_3._getErrorResponseParserFn(repository.context.registry));
}
function __requestRelations(repository, uri, requestOptions, queryData) {
    if (!repository.context.registry.inScope(uri, true))
        return Promise.reject(new IllegalArgumentError_1.IllegalArgumentError("\"" + uri + "\" is out of scope."));
    var url = repository.context
        .getObjectSchema()
        .resolveURI(uri, { base: true });
    var queryContainer = new QueryContainer_1.QueryContainer(repository.context, {
        containerPropertyType: queryData.containerPropertyType,
        uri: url,
    });
    __executeQueryBuilder(queryContainer, queryData);
    return __requestQueryDocuments(repository, url, requestOptions, queryContainer);
}
function __requestDocuments(repository, uris, requestOptions, queryData) {
    for (var _i = 0, uris_1 = uris; _i < uris_1.length; _i++) {
        var uri = uris_1[_i];
        if (!repository.context.registry.inScope(uri, true))
            return Promise.reject(new IllegalArgumentError_1.IllegalArgumentError("\"" + uri + "\" is out of scope."));
    }
    var urls = uris.map(function (uri) { return repository.context
        .getObjectSchema()
        .resolveURI(uri, { base: true }); });
    var queryContainer = new QueryContainer_1.QueryContainer(repository.context, {
        uris: urls,
    });
    __executeQueryBuilder(queryContainer, queryData);
    var url = urls.length === 1 ? urls[0] : repository.context.baseURI;
    Request_1.RequestUtils.setRetrievalPreferences({ include: [C_1.C.PreferDocumentChecksums] }, requestOptions);
    return __requestQueryDocuments(repository, url, requestOptions, queryContainer);
}
function __getQueryable(repository, uri, requestOptions, queryBuilderFn) {
    return __requestDocuments(repository, [uri], requestOptions, { queryBuilderFn: queryBuilderFn })
        .then(function (documents) { return documents[0]; });
}
function __addRefreshProperties(parentProperty, queryableProperty) {
    queryableProperty.subProperties.forEach(function (subProperty, propertyName) {
        var queryProperty = parentProperty._addSubProperty(propertyName, subProperty);
        __addRefreshProperties(queryProperty, subProperty);
    });
}
function __refreshQueryable(repository, document, requestOptions) {
    if (requestOptions === void 0) { requestOptions = {}; }
    if (!repository.context.registry.inScope(document.$id, true))
        return Promise.reject(new IllegalArgumentError_1.IllegalArgumentError("\"" + document.$id + "\" is out of scope."));
    var url = repository.context
        .getObjectSchema()
        .resolveURI(document.$id, { base: true });
    var queryContainer = new QueryContainer_1.QueryContainer(repository.context, { uris: [url] });
    __addRefreshProperties(queryContainer._queryProperty, document.$_queryableMetadata);
    Request_1.RequestUtils.setRetrievalPreferences({ include: [C_1.C.PreferDocumentChecksums] }, requestOptions);
    return __requestQueryDocuments(repository, url, requestOptions, queryContainer)
        .then(function (documents) { return documents[0]; });
}
exports.QueryableDocumentsRepositoryTrait = {
    PROTOTYPE: {
        get: function (uriOrURIs, requestOptionsOrQueryBuilderFn, queryBuilderFn) {
            var requestOptions = typeof requestOptionsOrQueryBuilderFn === "object" ?
                requestOptionsOrQueryBuilderFn : {};
            queryBuilderFn = Utils_2.isFunction(requestOptionsOrQueryBuilderFn) ?
                requestOptionsOrQueryBuilderFn : queryBuilderFn;
            if (typeof uriOrURIs !== "string") {
                return __requestDocuments(this, uriOrURIs, requestOptions, {
                    rootType: queryBuilderFn ? void 0 : QueryPropertyType_1.QueryPropertyType.FULL,
                    queryBuilderFn: queryBuilderFn,
                });
            }
            var uri = uriOrURIs;
            var target = this.context.registry.hasPointer(uri) ?
                this.context.registry.getPointer(uri, true) :
                void 0;
            if (queryBuilderFn) {
                var types_1 = target ? target.types : [];
                return __getQueryable(this, uri, requestOptions, function (_) {
                    types_1.forEach(function (type) { return _.withType(type); });
                    return queryBuilderFn.call(void 0, _);
                });
            }
            if (target && target.$isQueried())
                requestOptions.ensureLatest = true;
            return LDPDocumentsRepositoryTrait_1.LDPDocumentsRepositoryTrait.PROTOTYPE
                .get.call(this, uri, requestOptions)
                .then(function (document) {
                if (!document.$_queryableMetadata)
                    return document;
                var resources = document.$getFragments();
                resources.push(document);
                resources.forEach(function (resource) {
                    resource.$_queryableMetadata = void 0;
                });
                return document;
            });
        },
        resolve: function (document, requestOptionsOrQueryBuilderFn, queryBuilderFn) {
            return this.get(document.$id, requestOptionsOrQueryBuilderFn, queryBuilderFn);
        },
        refresh: function (document, requestOptions) {
            if (!document.$isQueried())
                return LDPDocumentsRepositoryTrait_1.LDPDocumentsRepositoryTrait.PROTOTYPE
                    .refresh.call(this, document, requestOptions);
            return __refreshQueryable(this, document, requestOptions);
        },
        saveAndRefresh: function (document, requestOptions) {
            var _this = this;
            if (!document.$_queryableMetadata)
                return LDPDocumentsRepositoryTrait_1.LDPDocumentsRepositoryTrait.PROTOTYPE
                    .saveAndRefresh.call(this, document, requestOptions);
            if (document.$eTag === null)
                return Promise.reject(new IllegalStateError_1.IllegalStateError("The document \"" + document.$id + "\" is locally outdated and cannot be saved."));
            var cloneOptions = Request_1.RequestUtils.cloneOptions(requestOptions || {});
            return this.save(document, cloneOptions)
                .then(function (doc) {
                return __refreshQueryable(_this, doc, requestOptions);
            });
        },
        getChildren: function (uri, requestOptionsOrQueryBuilderFn, queryBuilderFn) {
            var requestOptions = typeof requestOptionsOrQueryBuilderFn === "object" ?
                requestOptionsOrQueryBuilderFn : {};
            queryBuilderFn = Utils_2.isFunction(requestOptionsOrQueryBuilderFn) ?
                requestOptionsOrQueryBuilderFn : queryBuilderFn;
            Request_1.RequestUtils.setRetrievalPreferences({ include: [C_1.C.PreferDocumentChecksums] }, requestOptions);
            return __requestRelations(this, uri, requestOptions, {
                rootType: queryBuilderFn ? void 0 : QueryPropertyType_1.QueryPropertyType.FULL,
                containerPropertyType: QueryContainerPropertyType_1.QueryContainerPropertyType.CHILD,
                queryBuilderFn: queryBuilderFn,
            });
        },
        getMembers: function (uri, requestOptionsOrQueryBuilderFn, queryBuilderFn) {
            var requestOptions = typeof requestOptionsOrQueryBuilderFn === "object" ?
                requestOptionsOrQueryBuilderFn : {};
            queryBuilderFn = Utils_2.isFunction(requestOptionsOrQueryBuilderFn) ?
                requestOptionsOrQueryBuilderFn : queryBuilderFn;
            Request_1.RequestUtils.setRetrievalPreferences({ include: [C_1.C.PreferDocumentChecksums] }, requestOptions);
            return __requestRelations(this, uri, requestOptions, {
                rootType: queryBuilderFn ? void 0 : QueryPropertyType_1.QueryPropertyType.FULL,
                containerPropertyType: QueryContainerPropertyType_1.QueryContainerPropertyType.MEMBER,
                queryBuilderFn: queryBuilderFn,
            });
        },
        listChildren: function (uri, requestOptions) {
            if (requestOptions === void 0) { requestOptions = {}; }
            return __requestRelations(this, uri, requestOptions, {
                containerPropertyType: QueryContainerPropertyType_1.QueryContainerPropertyType.CHILD,
                rootType: QueryPropertyType_1.QueryPropertyType.EMPTY,
            });
        },
        listMembers: function (uri, requestOptions) {
            if (requestOptions === void 0) { requestOptions = {}; }
            return __requestRelations(this, uri, requestOptions, {
                rootType: QueryPropertyType_1.QueryPropertyType.EMPTY,
                containerPropertyType: QueryContainerPropertyType_1.QueryContainerPropertyType.MEMBER,
            });
        },
    },
    isDecorated: function (object) {
        return ModelDecorator_1.ModelDecorator
            .hasPropertiesFrom(exports.QueryableDocumentsRepositoryTrait.PROTOTYPE, object);
    },
    decorate: function (object) {
        if (exports.QueryableDocumentsRepositoryTrait.isDecorated(object))
            return object;
        var target = ModelDecorator_1.ModelDecorator
            .decorateMultiple(object, LDPDocumentsRepositoryTrait_1.LDPDocumentsRepositoryTrait);
        return ModelDecorator_1.ModelDecorator
            .definePropertiesFrom(exports.QueryableDocumentsRepositoryTrait.PROTOTYPE, target);
    },
};
