// carbonldp/.esm2015/RDF/Document.js
import * as Utils from "./../Utils";
import { RDFNode } from "./Node";
import { URI } from "./URI";
export const RDFDocument = {
    is(value) {
        return Utils.hasProperty(value, "@graph")
            && Utils.isArray(value["@graph"]);
    },
    create(resources, uri) {
        return {
            "@id": uri ? uri : "",
            "@graph": resources,
        };
    },
    getDocuments(objects) {
        if (Utils.isArray(objects))
            return objects
                .filter(RDFDocument.is);
        if (RDFDocument.is(objects))
            return [objects];
        return [];
    },
    getFreeNodes(objects) {
        if (!Array.isArray(objects))
            return [];
        return objects
            .filter(element => !RDFDocument.is(element))
            .filter(RDFNode.is);
    },
    getResources(objects) {
        const resources = RDFDocument.getFreeNodes(objects);
        RDFDocument
            .getDocuments(objects)
            .map(document => document["@graph"])
            .forEach(nodes => resources.push(...nodes));
        return resources;
    },
    getDocumentResources(document) {
        return RDFDocument
            .getResources(document)
            .filter(node => !RDFNode.isFragment(node));
    },
    getNamedFragmentResources(document, documentResource) {
        const uriToMatch = Utils.isObject(documentResource) ?
            RDFNode.getID(documentResource) :
            documentResource;
        return RDFDocument
            .getResources(document)
            .filter(node => {
            const id = RDFNode.getID(node);
            if (!URI.hasFragment(id))
                return;
            if (!uriToMatch)
                return true;
            return URI.getDocumentURI(id) === uriToMatch;
        });
    },
    getBNodeResources(document) {
        return RDFDocument
            .getResources(document)
            .filter(node => {
            const id = RDFNode.getID(node);
            return URI.isBNodeID(id);
        });
    },
    getNodes(rdfDocument) {
        const documentNodes = [];
        const fragmentNodes = [];
        for (const node of rdfDocument["@graph"]) {
            (RDFNode.isFragment(node) ? fragmentNodes : documentNodes).push(node);
        }
        return [documentNodes, fragmentNodes];
    },
};
